lib/cli/src/commands/run.rs

```diff
@@ -92,94 +90,154 @@ fn is_dir(e: &walkdir::DirEntry) -> bool {
-        command: Option<&str>,
-        _debug_output_allowed: bool,
-    ) -> Result<Run, anyhow::Error> {
-        let (manifest, pathbuf) =
-            wasmer_registry::get_executable_file_from_path(&package_root_dir, command)?;
+#[derive(Debug, Parser, Clone, Default)]
+/// The options for the `wasmer run` subcommand
+pub struct Run {
+    /// File, URL or package to run
+    #[clap(name = "FILE", parse(try_from_str))]
+    pub(crate) path: SplitVersion,
 
-        #[cfg(feature = "wasi")]
-        {
-            let default = HashMap::default();
-            let fs = manifest.fs.as_ref().unwrap_or(&default);
-            for (alias, real_dir) in fs.iter() {
-                let real_dir = package_root_dir.join(&real_dir);
-                if !real_dir.exists() {
-                    if _debug_output_allowed {
-                        println!(
-                            "warning: cannot map {alias:?} to {}: directory does not exist",
-                            real_dir.display()
-                        );
-                    }
-                    continue;
-                }
+    /// Registry to run the FILE from (optional)
+    #[clap(short = 'r', long, name = "registry")]
+    pub(crate) registry: Option<String>,
 
-                self.wasi.map_dir(alias, real_dir.clone());
-            }
-        }
+    #[clap(flatten)]
+    pub(crate) options: RunWithoutFile,
+}
+
+impl Run {
+    /// Resolves the `self.path` to an actual file path by downloading / querying the registry if necessary
+    /// Will block the current thread when doing so.
+    pub fn resolve(&self) -> Result<RunWithPathBuf, anyhow::Error> {
+        #[cfg(not(feature = "debug"))]
+        let debug = false;
+        #[cfg(feature = "debug")]
+        let debug = self.options.debug;
+        self.path
+            .get_run_command(self.registry.as_deref(), self.options.clone(), debug)
+    }
+
+    /// Create Run instance for arguments/env,
+    /// assuming we're being run from a CFP binfmt interpreter.
+    pub fn from_binfmt_args() -> Run {
+        Self::from_binfmt_args_fallible().unwrap_or_else(|e| {
+            crate::error::PrettyError::report::<()>(
+                Err(e).context("Failed to set up wasmer binfmt invocation"),
+            )
+        })
+    }
+
+    #[cfg(target_os = "linux")]
+    fn from_binfmt_args_fallible() -> Result<Run> {
+        use regex::Split;
 
-        Ok(Run {
-            path: pathbuf,
+        let argv = std::env::args_os().collect::<Vec<_>>();
+        let (_interpreter, executable, original_executable, args) = match &argv[..] {
+            [a, b, c, d @ ..] => (a, b, c, d),
+            _ => {
+                bail!("Wasmer binfmt interpreter needs at least three arguments (including $0) - must be registered as binfmt interpreter with the CFP flags. (Got arguments: {:?})", argv);
+            }
+        };
+        // TODO: Optimally, args and env would be passed as an UTF-8 Vec.
```

Comment:
why not just use `std::env::args()`, which gives a `String` iterator?

Reply:
I used `std::env::args()` now, not sure why it used `args_os()` originally.